{"pageProps":{"post":{"title":"Memoizing async functions so that you don't cache errors","date":"2022-02-26","slug":"2022-02-26-memoize-async","html":"<p>There are two hard problems in computer science: <a href=\"https://martinfowler.com/bliki/TwoHardThings.html\">Cache invalidation and naming\nthings</a>. In this post we'll\nshow how memoize an async function, and how to invalidate the memoization when\nthe promise throws an error.</p>\n<p>This helps us with being able to re-try because since the error is not cached,\ncalling it again after an error retries automatically.</p>\n<p>Example async function: fetch from the pokemon API</p>\n<pre><code class=\"language-javascript\">async function getPokemon() {\n  const id = Math.floor(Math.random() * 150)\n  const url = 'https://pokeapi.co/api/v2/pokemon/' + id\n  const ret = await fetch(url)\n  if (!ret.ok) {\n    throw new Error(\n      `Failed to fetch ${url} HTTP ${ret.status} ${ret.statusText}`,\n    )\n  }\n  return ret.json()\n}\n</code></pre>\n<p>Here is a technique that can be used to memoize this function</p>\n<pre><code class=\"language-javascript\">function getPokemonMemoized() {\n  if (!this.promise) {\n    this.promise = getPokemon().catch(e => {\n      this.promise = undefined\n      throw e\n    })\n  }\n  return this.promise\n}\n</code></pre>\n<p>The promise is held in this.promise, and the important part of this function is\nthat when I get an error, I clear this.promise and re-throw the error. The caller\nof the function, on error, will receive the error message, but caching will not\ntake place, allowing retries to take place later on.</p>\n<p>See <a href=\"https://cmdcolin.github.io/pokemon.html\">https://cmdcolin.github.io/pokemon.html</a> for demo</p>\n<h2>Footnote 0: Arguments to function</h2>\n<p>If your function takes arguments, then you can use a hashmap associating the\nargument with the promise. You may also consider using an LRU cache so that\nyour hashmap doesn't grow infinitely in size</p>\n<p>Generally you need a way to stringify or otherwise make them able to be stored\nin a Map or Object to do this.</p>\n<p>See <a href=\"https://github.com/nodeca/promise-memoize\">https://github.com/nodeca/promise-memoize</a> for example</p>\n<h2>Footnote 1: Error handling of <code>fetch</code></h2>\n<p>This demo also demonstrates some basic fetch error handling, and uses\n<code>statusText</code> <a href=\"https://developer.mozilla.org/en-US/docs/Web/API/Response/statusText\">which happens to not exist in\nHTTP/2</a>.\nIf you want a semblence of status message text in HTTP/2 you can try to use\nawait ret.json() (if the API returns json error messages) or await ret.text()\ninside the catch clause, but note that it could cause yet another error to be\nthrown</p>\n<h2>Footnote 2: Global cache</h2>\n<p>You could also keep a cache in a global variable, or as a property on a class,\nor other methods. I have also found it useful to have a specific function for\nclearing the cache, so you can get a clean slate each time a test runs in unit\ntesting or similar</p>\n<pre><code class=\"language-javascript\">let promise\nasync function getPokemonMemoized() {\n  if (!promise) {\n    promise = getPokemon().catch(e => {\n      promise = undefined\n      throw e\n    })\n  }\n  return promise\n}\nfunction clearCache() {\n  promise = undefined\n}\n</code></pre>\n<p>You can also make a general purpose utility to memoize any promise function</p>\n<pre><code class=\"language-javascript\">function memoize(fn) {\n  let promise\n  return () => {\n    if (!promise) {\n      promise = fn().catch(e => {\n        promise = undefined\n        throw e\n      })\n    }\n  }\n}\n</code></pre>\n<h2>Footnote 3 - Aborting</h2>\n<p>If you want to handle aborting, it is a bit trickier. Aborting in javascript is\nhandled by\n<a href=\"https://developer.mozilla.org/en-US/docs/Web/API/AbortController/AbortController\">AbortController</a>.\nThis is an object that gives you an\n<a href=\"https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal\">AbortSignal</a>\nthat can be passed to fetch calls and the like to stop a big download from\nhappening.</p>\n<p>In our above example, if we passed an abort signal to the first call to fetch,\nand then aborted it, it would abort the fetch, <a href=\"https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort\">which throws a DOMException\ncalled\n\"AbortError\"</a>.\nYou can detect that it is an AbortError like this, and may choose not to\ndisplay or re-throw the abort exception</p>\n<pre><code class=\"language-javascript\">function isAbortException(e) {\n  return e instanceof Error &#x26;&#x26; exception.name === 'AbortError'\n}\n</code></pre>\n<p>Now, what if 5 functions call getPokemonMemoized(), all passing different abort\nsignals. What if the first one aborts? Then all the rest will get aborted also.\nBut what if we only want to abort the cached call if literally all of them\naborted? Then we may have to synthesize an abortcontroller inside our function</p>\n<pre><code class=\"language-javascript\">let promise\nlet abortcontroller\nlet listeners = 0\nasync function getPokemonMemoized(signal) {\n  if (!promise) {\n    abortcontroller = new AbortController()\n\n    // synthesize a new signal instead of using the passed in signal\n    promise = getPokemon(abortcontroller.signal).catch(e => {\n      promise = undefined\n      throw e\n    })\n  }\n  if (signal) {\n    listeners++\n    // add listener to the passed in signal\n    signal.addEventListener('abort', () => {\n      listeners--\n      if (listeners === 0) {\n        abortcontroller.abort()\n      }\n    })\n  }\n  return promise\n}\n</code></pre>\n<p>A library my team created,\n<a href=\"https://github.com/GMOD/abortable-promise-cache\">abortable-promise-cache</a>,\ntries to help with this scenario with a cleaner abstraction.</p>\n<h2>Footnote 4</h2>\n<p>I have been playing through Pokemon Yellow and find it really amusing hence the\npokemon theme</p>\n<p>Fun stuff: The cutting room floor wiki with unused moves, sounds, and sprites\nin Pokemon Yellow <a href=\"https://tcrf.net/Pok%C3%A9mon_Yellow\">https://tcrf.net/Pok%C3%A9mon_Yellow</a></p>\n<h2>Footnote 5</h2>\n<p>This blog post mentioned in a comment thread <a href=\"https://zansh.in/memoizer.html\">https://zansh.in/memoizer.html</a> has\ngreat interactive examples and shows the \"invalidate on .catch()\" behavior!</p>"}},"__N_SSG":true}